arduino-ide-extension/src/browser/boards/boards-service-provider.ts

```diff
@@ -205,20 +199,36 @@ export class BoardsServiceProvider
         Port.sameAs(board.port, port)
       );
 
-      const lastBoardsConfigOnUpload = this.lastBoardsConfigOnUpload!;
+      const lastBoardsConfigOnUpload = this.lastBoardsConfigOnUpload;
 
-      if (
-        boardOnAppearedPort &&
-        lastBoardsConfigOnUpload.selectedBoard &&
-        Board.sameAs(
+      if (boardOnAppearedPort && lastBoardsConfigOnUpload.selectedBoard) {
+        const boardIsSameHardware = Board.hardwareIdEquals(
           boardOnAppearedPort,
           lastBoardsConfigOnUpload.selectedBoard
-        )
-      ) {
+        );
+
+        const boardIsSameFqbn = Board.sameAs(
+          boardOnAppearedPort,
+          lastBoardsConfigOnUpload.selectedBoard
+        );
+
+        if (!boardIsSameHardware && !boardIsSameFqbn) return;
```

Comment:
Instead of `return`, it should be `continue`, right?

Reply:
this is the same as in [1885](https://github.com/arduino/arduino-ide/pull/1885), and should result in the same reasoning as 2.0.3, but considering `hardwareId`: if a board that "appeared" does not have the same `hardwareId` or `fqbn` as the one selected for upload, don't do anything and return.

Reply:
also to clarify this logic is only related to port change during upload

Reply:
No. Without this change, IDE2 goes through all appeared ports and tries to find the matching one. It's a noop, when a match was found. With your changes, the logic breaks after the first non-match.

Reply:
for context, in 2.0.3 we have [this](https://github.com/arduino/arduino-ide/blob/01ee045beb1cd69b23bc68c83b918bb14ab6ef91/arduino-ide-extension/src/browser/boards/boards-service-provider.ts#L210), the "if block" is avoided if the "appeared board" is not the same fqbn, so same as doing a return as above (but it also considers `hardwareId`)